apps/dot_ext/tests/test_authorization.py

```diff
@@ -92,7 +101,108 @@ def test_post_with_invalid_non_standard_scheme(self):
+            client_type=Application.CLIENT_CONFIDENTIAL,
+            redirect_uris=redirect_uri)
+        application.scope.add(capability_a, capability_b)
+
+        # user logs in
+        self.client.login(username='anna', password='123456')
+
+        code_challenge = "sZrievZsrYqxdnu2NVD603EiYBM18CuzZpwB-pOSZjo"
+
+        payload = {
+            'client_id': application.client_id,
+            'response_type': 'code',
+            'redirect_uri': redirect_uri,
+            'code_challenge': code_challenge,
+            'code_challenge_method': 'S256',
+        }
+        response = self.client.get('/v1/o/authorize', data=payload)
+        # post the authorization form with only one scope selected
+        payload = {
+            'client_id': application.client_id,
+            'response_type': 'code',
+            'redirect_uri': redirect_uri,
+            'scope': ['capability-a'],
+            'expires_in': 86400,
+            'allow': True,
+            'code_challenge': code_challenge,
+            'code_challenge_method': 'S256',
+        }
+        response = self.client.post(response['Location'], data=payload)
 
+        self.assertEqual(response.status_code, 302)
+        # now extract the authorization code and use it to request an access_token
+        query_dict = parse_qs(urlparse(response['Location']).query)
+        authorization_code = query_dict.pop('code')
+        token_request_data = {
+            'grant_type': 'authorization_code',
+            'code': authorization_code,
+            'redirect_uri': redirect_uri,
+            'client_id': application.client_id,
+        }
+        # Test that request is unauthorized WITH OUT the client_secret.
+        token_request_data.update({'code_verifier': 'test123456789123456789123456789123456789123456789'})
+        response = self.client.post(reverse('oauth2_provider:token'), data=token_request_data)
+        self.assertEqual(response.status_code, 401)
+
+        # Setup application's client_secret
+        token_request_data.update({'client_secret': application.client_secret})
+
+        # Test that using a BAD code_verifier has a bad request response
+        token_request_data.update({'code_verifier': 'test1234567bad9verifier23456789123456789123456789'})
+        response = self.client.post(reverse('oauth2_provider:token'), data=token_request_data)
+        self.assertEqual(response.status_code, 400)
+
+        # Test that request is successful WITH the client_secret and GOOD code_verifier
+        token_request_data.update({'code_verifier': 'test123456789123456789123456789123456789123456789'})
+        response = self.client.post(reverse('oauth2_provider:token'), data=token_request_data)
+        self.assertEqual(response.status_code, 200)
+
+        # Test 2nd access token request is unauthorized
+        response = self.client.post(reverse('oauth2_provider:token'), data=token_request_data)
+        self.assertEqual(response.status_code, 401)
```

Comment:
Is it worth adding this 2nd token request to the client type public test as well?

Reply:
> Is it worth adding this 2nd token request to the client type public test for consistency?

@njdister I just pushed up a commit with this addition. TY!